src/allmydata/web/common.py

```diff
@@ -713,8 +704,15 @@ def url_for_string(req, url_string):
         )
     return url
 
+T = TypeVar("T")
+
+@overload
```

Comment:
TIL

Reply:
So this is saying:

* If multiple is False, we return bytes
* If multiple is True, we return tuple of bytes

But then the final version allows returning random other things, regardless of value of multiple. So how does that work?

Reply:
You are close.  The key is that `Literal[False] doesn't mean "False".  It means a statically detectable False.  For example:

```
get_arg(..., multiple=False)
```

or

```
x = False
get_arg(..., multiple=x)
```

but probably not

```
get_arg(..., multiple=True if x else False)
```

I don't know the exact extent to which mypy can track this down.

So the first two cases are for statically detectable False or True and the third case covers those _and_ the case where the value couldn't be determined.

Writing this out and reconsidering it, I'm not sure how much sense it makes.  I think I got lost in making mypy happy and wasn't thinking carefully enough about the real meaning.  A couple things occur to me now:

* The `T` case is just as possible for the `Literal[True]` and `Literal[False]` cases as for the third case.
* The `None` in the return type seems like nonsense.  The only way to get None back is if that is the value given for `default`.  It seems like `T` should cover this but maybe it doesn't because of some special cases applied to `None`?

sigh... I'll see if I can work this into something more coherent.

Reply:
2490f0f58 fwiw